### ISSUE TITLE
Helm v3 - linting multiple charts depends on the order of the charts

### ISSUE THREAD

**Issue by andytom** [helm/helm#6079]
Output of `helm version`:
```
version.BuildInfo{Version:"v3.0.0-alpha.2", GitCommit:"97e7461e41455e58d89b4d7d192fed5352001d44", GitTreeState:"clean", GoVersion:"go1.12.7"}
```

Output of `kubectl version`:

N/A

Cloud Provider/Platform (AKS, GKE, Minikube etc.):  

N/A

----
When linting multiple charts using helm3 the order that the charts are passed in matters. The following was done using the, at time of writing, latest commit on the master branch of the charts repo https://github.com/helm/charts/commit/8fb57b472e7a9d5117734566146f272ebc071aa2.

```
$ helm3 lint stable/atlantis stable/apm-server
2 chart(s) linted, 0 chart(s) failed
```

```
$ helm3 lint stable/apm-server stable/atlantis
Error: 2 chart(s) linted, 1 chart(s) failed
	%!s(<nil>)
	[ERROR] templates/: render error at (atlantis/templates/configmap-config.yaml:13:27): wrong type for value; expected string; got map[string]interface {}
```

Using the latest version of helm 2 (v2.14.2) this works as expected
```
$ helm lint stable/apm-server stable/atlantis
==> Linting stable/apm-server
Lint OK

==> Linting stable/atlantis
Lint OK

2 chart(s) linted, no failures
```

```
$ helm lint stable/atlantis stable/apm-server
==> Linting stable/atlantis
Lint OK

==> Linting stable/apm-server
Lint OK

2 chart(s) linted, no failures
```

**Comment by waveywaves** [comment]
Hello Team ! I would like to start working on this bug. 
Anything to keep in mind before I get started ?

**Comment by karuppiah7890** [comment]
@waveywaves you can start off and comment here if you have any questions while implementing 😊

**Comment by waveywaves** [comment]
Cool @karuppiah7890 Thanks for that :)

**Comment by waveywaves** [comment]
After reproducing the issue I have been trying to debug this for the past few days.
```
➜  helm3test helm lint ./apm-server ./atlantis
Error: 2 chart(s) linted, 1 chart(s) failed
        %!s(<nil>)
        [ERROR] templates/: template: atlantis/templates/configmap-config.yaml:13:27: executing "atlantis/templates/configmap-config.yaml" at <4>: wrong type for value; expected string; got map[string]interface
{}

➜  helm3test helm lint ./atlantis ./apm-server
2 chart(s) linted, 0 chart(s) failed
```
@karuppiah7890 Seems lke the issue is around https://github.com/helm/helm/blob/dev-v3/pkg/action/lint.go#L61-L73

I checked if there is anything wrong within `lintChart` because of which this might be taking place but had no luck finding anything. After checking with the `2.14` branch I can tice that there are a few changes regarding changing the datatype of vals from `[]byte` to `map[string]inteface{}` which might be important as the error I am seeing wrt 
`[ERROR] templates/: template: atlantis/templates/configmap-config.yaml:13:27: executing "atlantis/templates/configmap-config.yaml" at <4>: wrong type for value; expected string; got map[string]interface` 
might give us some idea. I could not find where `<4>` or rather did not understand where to find it.

After checking out https://github.com/helm/helm/commit/2b81eea1e289da0c0aa1f2f1ca1dca705fa07f70#diff-a3b833d32c8c6c71b4cfd64cf3cdce79 Seems like the answer might lie here. I am a little lost right now and don't know if I am looking in the right direction exactly. Any kind of help would be appreciated.

**Comment by karuppiah7890** [comment]
@waveywaves I checked the issue with a fresh look, also I didn't entirely understand the things that you mentioned as I have never seen the linting code before. So I checked it out. I debugged it using the IDE then and saw the issue with `vals` that you mentioned. You are right about the change https://github.com/helm/helm/commit/2b81eea1e289da0c0aa1f2f1ca1dca705fa07f70#diff-a3b833d32c8c6c71b4cfd64cf3cdce79 being the issue . 

While debugging I noticed the default values of apm-server is getting mixed up with the deafult values of atlantis and then the templates of atlantis are rendered. 

On checking further, I realized that this happened because the `values` map over here : 

https://github.com/helm/helm/blob/68ee30b48cd0e34f2404f7ea56f491b34158edee/pkg/action/lint.go#L62

is used for linting both the charts and while linting the first one it gets mutated during the process and things get mixed up. You can see the mutation happen if you follow this - 

https://github.com/helm/helm/blob/68ee30b48cd0e34f2404f7ea56f491b34158edee/pkg/action/lint.go#L77

https://github.com/helm/helm/blob/68ee30b48cd0e34f2404f7ea56f491b34158edee/pkg/action/lint.go#L113

https://github.com/helm/helm/blob/895e9192d440e144ead0796fd8c1db81b47ef1c6/pkg/lint/lint.go#L34

https://github.com/helm/helm/blob/ec440d446d4ffebfaf0c4c52c4b0e06edb27cb25/pkg/lint/rules/template.go#L58

https://github.com/helm/helm/blob/b49db9e6e6fb1913a125b8cf9b31f81199c6753c/pkg/chartutil/coalesce.go#L40

https://github.com/helm/helm/blob/b49db9e6e6fb1913a125b8cf9b31f81199c6753c/pkg/chartutil/coalesce.go#L50

in which `v` gets mutated. Mutation happens because maps are passed by reference I think, didn't check more on that. Could you check this out? I think the whole issue could be solved if we do a clone of `vals` over here 

https://github.com/helm/helm/blob/68ee30b48cd0e34f2404f7ea56f491b34158edee/pkg/action/lint.go#L62

and then send the clone to the `lintChart` function, so that the clone gets mutated, but doesn't affect the next loop, and the values set in the command line, if any, will also be used up for linting all the charts

Let me know if you need any help on this. We could even remote pair on zoom may be. Also, I was thinking it would be great to TDD - first write a failing test, and then fix the issue to pass the test. I have always tried to do it. It helps keep the features more robust 😄

**Comment by bacongobbler** [comment]
> Also, I was thinking it would be great to TDD - first write a failing test, and then fix the issue to pass the test. I have always tried to do it. It helps keep the features more robust smile

Yeah, it's always best to cover issues with a unit test when possible. Certain parts of the codebase are notoriously hard to unit test (`pkg/kube` for example), but `pkg/chartutil` should be easy to write a test for.

**Comment by karuppiah7890** [comment]
@waveywaves Were you able to work on this?

**Comment by waveywaves** [comment]
@karuppiah7890 I figured it out some time ago only. Got really confused while debugging this. Will continue with this and let yu know if I need any help. Thank you so much :)

**Comment by andytom** [comment]
Looks like this was fixed by #6241.
